extensions/wikia/UserLogin/UserLoginHelper.class.php

```diff
@@ -133,97 +133,83 @@ protected function addUserToUserList( $userId, &$userList ) {
 	}
 
 	/**
-	 * get a list of random popular wikis
-	 * @param integer $require (number of wikis)
-	 * @return array $wikis
+	 * Redirect the user to the appropriate page after login.
+	 *
+	 * @param string $extraReturnToQuery Use this to add any additional parameters to the query string
+	 * @param int $cbVal Send a specific CB value (mostly for testing)
 	 */
-	public function getRandomWikis( $require=self::LIMIT_WIKIS ) {
-		return array();
-/**
-		$memKey = wfSharedMemcKey( 'userlogin', 'random_wikis' );
-		$wikis = $this->wg->Memc->get( $memKey );
-		if ( !is_array($wikis) ) {
-			$popularWikis = $this->getPopularWikis();
-			shuffle( $popularWikis );
-			$wikis = array();
-
-			foreach( $popularWikis as $wikiId ) {
-				$themeSettings = WikiFactory::getVarValueByName( 'wgOasisThemeSettings', $wikiId);
-				if( !empty($themeSettings['wordmark-image-url']) ) {
-					$wikis[] = $themeSettings['wordmark-image-url'];
-				}
-
-				if ( count($wikis) >= $require )
-					break;
-			}
-
-			$this->wg->Memc->set( $memKey, $wikis, 60*60*24 );
-		}
-
-		return $wikis;
- **/
+	public function doRedirect( $extraReturnToQuery = '', $cbVal = 0 ) {
+		$this->wg->out->redirect( $this->getRedirectUrl( $extraReturnToQuery, $cbVal ) );
 	}
 
 	/**
-	 * get popular wikis
-	 * @param integer $wikiId
-	 * @param integer $limit (number of users)
-	 * @return array $popularWikis
+	 * Determine the URL that should be redirected to after the current login action.  If
+	 * there is a returnto URL and its valid then send them there.  Otherwise send the
+	 * user to the wikia's main page.
+	 *
+	 * @param string $extraReturnToQuery Use this to add any additional parameters to the query string
+	 * @param int $cbVal Send a specific CB value (mostly for testing)
+	 *
+	 * @return String
+	 * @throws MWException
 	 */
-	protected function getPopularWikis( $limit=50 ) {
-		wfProfileIn( __METHOD__ );
+	public function getRedirectUrl( $extraReturnToQuery = '', $cbVal = 0 ) {
+		$returnUrl = $this->wg->Request->getVal( 'returnto', '' );
+		$titleObj = Title::newFromText( $returnUrl );
 
-		$memKey = wfSharedMemcKey( 'userlogin', 'popular_wikis' );
-		$popularWikis = $this->wg->Memc->get( $memKey );
-		if ( empty($popularWikis) ) {
-			$popularWikis = array_keys( DataMartService::getTopWikisByPageviews( DataMartService::PERIOD_ID_MONTHLY, $limit ) );
+		if ( $this->isInvalidLoginRedirect( $titleObj ) ) {
+			$titleObj = Title::newMainPage();
+		}
 
-			if ( empty($popularWikis) ) {
-				$popularWikis[] = self::WIKIA_CITYID_COMMUNITY;
-			}
+		$cbParam = 'cb=' . ( $cbVal ? $cbVal : rand( 1, 10000 ) );
+		$returnParams = $this->wg->Request->getVal( 'returntoquery', '' );
 
-			$this->wg->Memc->set( $memKey, $popularWikis, 60*60*24 );
+		if ( !empty( $extraReturnToQuery ) ) {
+			$returnParams .= ( empty( $returnParams ) ? '' : '&' ) . $extraReturnToQuery;
```

Comment:
It looks like `$titleObj->getFullURL` takes an array as well as a string. I think an array would look cleaner, however it might not be a convenient if `$returnParams` is already set as a string. Just putting it out there in case you have any thoughts on that, @garthwebb.